tensorflow/core/kernels/depthtospace_op.cc

```diff
@@ -142,34 +143,71 @@ class DepthToSpaceOp : public OpKernel {
   TensorFormat data_format_;
 };
 
+namespace generator {
+template <typename T>
+struct DepthToSpaceGenerator {
+  EIGEN_ALWAYS_INLINE DepthToSpaceGenerator(
+      typename TTypes<T, 4>::ConstTensor input, const Eigen::Index block_size,
+      const Eigen::Index output_height, const Eigen::Index output_width,
+      const Eigen::Index output_depth)
+      : input_(input),
+        block_size_(block_size),
+        output_height_(output_height),
+        output_width_(output_width),
+        output_depth_(output_depth),
+        ys_(output_height),
+        xs_(output_width) {
+    compute_indices(ys_);
+    compute_indices(xs_);
+  }
+
+  EIGEN_ALWAYS_INLINE T
+  operator()(const Eigen::array<Eigen::Index, 4>& coords) const {
+    const Eigen::Index b = coords[0];
+    const Eigen::Index h = coords[1];
+    const Eigen::Index w = coords[2];
+    const Eigen::Index d = coords[3];
+    const Eigen::Index in_h = ys_[h].first;
+    const Eigen::Index offset_h = ys_[h].second;
+    const Eigen::Index in_w = xs_[w].first;
+    const Eigen::Index offset_w = xs_[w].second;
+    const Eigen::Index offset_d =
+        (offset_h * block_size_ + offset_w) * output_depth_;
+    const Eigen::Index in_d = d + offset_d;
+    return input_(b, in_h, in_w, in_d);
+  }
+
+  EIGEN_ALWAYS_INLINE void compute_indices(
+      std::vector<std::pair<Eigen::Index, Eigen::Index>>& indices) {
+    // Cache input indices to reduce repetitive computation.
+    for (Eigen::Index i = 0; i < indices.size(); ++i) {
+      indices[i].first = i / block_size_;
+      indices[i].second = i % block_size_;
+    }
+  }
+
+  typename TTypes<T, 4>::ConstTensor input_;
+  const Eigen::Index block_size;
+  const Eigen::Index output_height;
+  const Eigen::Index output_width;
+  const Eigen::Index output_depth;
+  std::vector<std::pair<Eigen::Index, Eigen::Index>> ys_, xs_;
```

Comment:
Technically you only need one of these, sized to the larger of `output_width` and `output_height`.